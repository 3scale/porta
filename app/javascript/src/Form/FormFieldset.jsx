// @flow
// TODO: Replace this component when patternfly-react implements it.

import * as React from 'react'
// $FlowFixMe Flow has troubles with @patternfly modules
import { FormContext } from '@patternfly/react-core/dist/js/components/Form/FormContext'
// $FlowFixMe Flow has troubles with @patternfly modules
import styles from '@patternfly/react-styles/css/components/Form/form'
// $FlowFixMe Flow has troubles with @patternfly modules
import { css } from '@patternfly/react-styles'

type Props = {
  children?: React.Node,
  className?: string,
  isInline?: boolean
}

const FormFieldset = ({
  children,
  className = '',
  isInline = false,
  ...props
}: Props): React.Node => (
  <FormContext.Consumer>
    {({ isHorizontal }: { isHorizontal: boolean }) => (
      <fieldset
        {...props}
        className={css(styles.formFieldset, isInline ? styles.modifiers.inline : className)}
      >
        {isHorizontal ? <div className={css(styles.formHorizontalGroup)}>{children}</div> : children}
      </fieldset>
    )}
  </FormContext.Consumer>
)

export {
  FormFieldset
}
