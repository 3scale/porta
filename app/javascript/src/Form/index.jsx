// @flow

export * from 'Form/FormFieldset'
export * from 'Form/FormLegend'
