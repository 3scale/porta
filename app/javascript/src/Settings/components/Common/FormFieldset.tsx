// TODO: Replace this component when patternfly-react implements it.

import * as React from 'react'

import styles from '@patternfly/react-styles/css/components/Form/form'
import { css } from '@patternfly/react-styles'

type Props = React.FieldsetHTMLAttributes<HTMLFieldSetElement> & {
  isInline?: boolean
}

const FormFieldset: React.FunctionComponent<Props> = ({ children, className = '', isInline = false, ...props }) => (
  <fieldset {...props} className={css(styles.formFieldset, isInline ? styles.modifiers.inline : className)} >
    {children}
  </fieldset>
)

export { FormFieldset }
