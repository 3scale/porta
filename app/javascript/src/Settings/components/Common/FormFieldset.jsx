// @flow
// TODO: Replace this component when patternfly-react implements it.

import * as React from 'react'

import { FormContext } from '@patternfly/react-core/dist/js/components/Form/FormContext'
// $FlowFixMe[cannot-resolve-module] to fix this, try import via scss
import styles from '@patternfly/react-styles/css/components/Form/form'
import { css } from '@patternfly/react-styles'

type Props = {
  children?: React.Node,
  className?: string,
  isInline?: boolean
}

const FormFieldset = ({
  children,
  className = '',
  isInline = false,
  ...props
}: Props): React.Node => (
  <FormContext.Consumer>
    {({ isHorizontal }: { isHorizontal: boolean }) => (
      <fieldset
        {...props}
        className={css(styles.formFieldset, isInline ? styles.modifiers.inline : className)}
      >
        {isHorizontal ? <div className={css(styles.formHorizontalGroup)}>{children}</div> : children}
      </fieldset>
    )}
  </FormContext.Consumer>
)

export {
  FormFieldset
}
