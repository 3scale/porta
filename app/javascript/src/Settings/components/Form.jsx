// @flow

import * as React from 'react'
import { useState } from 'react'
import { FormFieldset, FormLegend } from 'Form'
import { FormCollection, TextInputGroup, RadioFieldset } from 'Settings/components/Common'
import { AuthenticationSettingsFieldset } from 'Settings/components/AuthenticationSettingsFieldset'
import type { SettingsProps as Props } from 'Settings/types'

const SERVICE_MESH_INTEGRATION = 'service_mesh_istio'
const PROXY_HOSTED_INTEGRATION = 'hosted'

const Form = ({
  isProxyCustomUrlActive,
  integrationMethod,
  authenticationMethod,
  proxyEndpoints,
  authenticationSettings,
  credentialsLocation,
  security,
  gatewayResponse
}: Props): React.Node => {
  const [ selectedIntegrationMethod, setSelectedIntegrationMethod ] = useState(integrationMethod.value)
  const [ selectedAuthenticationMethod, setSelectedAuthenticationMethod ] = useState(authenticationMethod.value)
  const onChange = (setState) => (_checked, e) => setState(e.currentTarget.value)
  const isServiceMesh = selectedIntegrationMethod === SERVICE_MESH_INTEGRATION
  const isProxyHosted = selectedIntegrationMethod === PROXY_HOSTED_INTEGRATION
  const isProxyUrlsReadOnly = !isProxyCustomUrlActive && isProxyHosted
  const customProxyEndpoints = proxyEndpoints.map(endpoint =>
    ({ ...endpoint, readOnly: isProxyUrlsReadOnly, isDefaultValue: isProxyUrlsReadOnly }))

  return (
    <React.Fragment>
      <RadioFieldset {...integrationMethod} onChange={onChange(setSelectedIntegrationMethod)} value={selectedIntegrationMethod} legend='Integration' />
      { !isServiceMesh && <FormCollection collection={customProxyEndpoints} ItemComponent={TextInputGroup} legend='API gateway' /> }
      <RadioFieldset {...authenticationMethod} onChange={onChange(setSelectedAuthenticationMethod)} value={selectedAuthenticationMethod} legend='Authentication' />
      {/* $FlowIgnore[cannot-spread-inexact] */}
      <AuthenticationSettingsFieldset
        isServiceMesh={isServiceMesh}
        authenticationMethod={selectedAuthenticationMethod}
        {...authenticationSettings}
      />
      { !isServiceMesh && <React.Fragment>
        <RadioFieldset {...credentialsLocation} legend='Credentials Location' />
        <FormCollection collection={security} ItemComponent={TextInputGroup} legend='Security' />
        <FormFieldset id='fieldset-GatewayResponse'>
          <FormLegend>Gateway Response</FormLegend>
          {gatewayResponse.map(settings => (
            <FormCollection key={settings.legend} collection={settings.collection} ItemComponent={TextInputGroup} legend={settings.legend} />
          ))}
        </FormFieldset>
      </React.Fragment> }
    </React.Fragment>
  )
}

export {
  Form
}
