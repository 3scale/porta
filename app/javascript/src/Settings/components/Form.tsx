import { useState } from 'react'

import {
  FormCollection,
  FormFieldset,
  FormLegend,
  TextInputGroup,
  RadioFieldset
} from 'Settings/components/Common'
import { AuthenticationSettingsFieldset } from 'Settings/components/AuthenticationSettingsFieldset'
import { createReactWrapper } from 'utilities/createReactWrapper'
import { SETTINGS_DEFAULT } from 'Settings/defaults'
import { FieldCatalogProps, FieldGroupProps, LegendCollectionProps, TypeItemProps } from 'Settings/types'

const SERVICE_MESH_INTEGRATION = 'service_mesh_istio'
const PROXY_HOSTED_INTEGRATION = 'hosted'

type Props = {
  isProxyCustomUrlActive: boolean,
  integrationMethod: FieldGroupProps & FieldCatalogProps,
  authenticationMethod: FieldGroupProps & FieldCatalogProps,
  proxyEndpoints: FieldGroupProps[],
  authenticationSettings: {
    appIdKeyPairSettings: FieldGroupProps[],
    apiKeySettings: FieldGroupProps,
    oidcSettings: {
      basicSettings: TypeItemProps,
      flowSettings: FieldGroupProps[],
      jwtSettings: TypeItemProps
    }
  },
  credentialsLocation: FieldGroupProps & FieldCatalogProps,
  security: FieldGroupProps[],
  gatewayResponse: LegendCollectionProps[]
}

const Form: React.FunctionComponent<Props> = ({
  isProxyCustomUrlActive,
  integrationMethod,
  authenticationMethod,
  proxyEndpoints,
  authenticationSettings,
  credentialsLocation,
  security,
  gatewayResponse
}) => {
  const [ selectedIntegrationMethod, setSelectedIntegrationMethod ] = useState(integrationMethod.value)
  const [ selectedAuthenticationMethod, setSelectedAuthenticationMethod ] = useState(authenticationMethod.value)
  const onChange = (setState: (arg1: ((arg1: string) => string) | string) => void) => (_checked: string, e: React.SyntheticEvent<HTMLButtonElement>) => setState(e.currentTarget.value)
  const isServiceMesh = selectedIntegrationMethod === SERVICE_MESH_INTEGRATION
  const isProxyHosted = selectedIntegrationMethod === PROXY_HOSTED_INTEGRATION
  const isProxyUrlsReadOnly = !isProxyCustomUrlActive && isProxyHosted
  const customProxyEndpoints = proxyEndpoints.map(endpoint => ({
    ...endpoint,
    readOnly: isProxyUrlsReadOnly,
    isDefaultValue: isProxyUrlsReadOnly
  }))

  return (
     <>
      <RadioFieldset {...integrationMethod} onChange={onChange(setSelectedIntegrationMethod)} value={selectedIntegrationMethod} legend='Integration' />
      { !isServiceMesh && <FormCollection collection={customProxyEndpoints} ItemComponent={TextInputGroup} legend='API gateway' /> }
      <RadioFieldset {...authenticationMethod} onChange={onChange(setSelectedAuthenticationMethod)} value={selectedAuthenticationMethod} legend='Authentication' />
      {/* $FlowIgnore[cannot-spread-inexact] */}
      <AuthenticationSettingsFieldset
        isServiceMesh={isServiceMesh}
        authenticationMethod={selectedAuthenticationMethod}
        {...authenticationSettings}
      />
      { !isServiceMesh &&  <>
        <RadioFieldset {...credentialsLocation} legend='Credentials Location' />
        <FormCollection collection={security} ItemComponent={TextInputGroup} legend='Security' />
        <FormFieldset id='fieldset-GatewayResponse'>
          <FormLegend>Gateway Response</FormLegend>
          {gatewayResponse.map(settings => (
            <FormCollection key={settings.legend} collection={settings.collection} ItemComponent={TextInputGroup} legend={settings.legend} />
          ))}
        </FormFieldset>
       </> }
     </>
  )
}

const FormWrapper = (settings: Props = SETTINGS_DEFAULT, elementId: string) => createReactWrapper(<Form {...settings} />, elementId)

export { Form, FormWrapper, Props }
