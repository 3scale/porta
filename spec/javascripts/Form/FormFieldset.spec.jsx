// @flow

import React from 'react'
import { mount } from 'enzyme'
import { FormFieldset } from 'Form/FormFieldset'

describe('FormFieldset', () => {
  it('should render default form fieldset variant', () => {
    const view = mount(
      <FormFieldset>
        <input id="input-id" />
      </FormFieldset>
    )
    expect(view).toMatchSnapshot()
  })

  it('should render inline form fieldset variant', () => {
    const view = mount(
      <FormFieldset isInline>
        <input id="input-id" />
      </FormFieldset>
    )
    expect(view).toMatchSnapshot()
  })

  it('should render form fieldset with custom class names', () => {
    const view = mount(
      <FormFieldset className='extra-class another-class'>
        <input id="input-id" />
      </FormFieldset>
    )
    expect(view).toMatchSnapshot()
  })
})
