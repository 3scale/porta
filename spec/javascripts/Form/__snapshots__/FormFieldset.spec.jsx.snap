// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFieldset should render default form fieldset variant 1`] = `
<FormFieldset
  fieldId="label-id"
>
  <fieldset
    className="pf-c-form__fieldset"
  >
    <input
      id="input-id"
    />
  </fieldset>
</FormFieldset>
`;

exports[`FormFieldset should render form fieldset invalid variant 1`] = `
<FormFieldset
  fieldId="label-id"
  helperTextInvalid="Invalid FormFieldset"
  isValid={false}
  label="label"
>
  <fieldset
    className="pf-c-form__fieldset"
  >
    <label
      className="pf-c-form__label"
      htmlFor="label-id"
    >
      <span
        className="pf-c-form__label-text"
      >
        label
      </span>
    </label>
    <input
      id="input-id"
    />
    <div
      aria-live="polite"
      className="pf-c-form__helper-text pf-m-error"
      id="label-id-helper"
    >
      Invalid FormFieldset
    </div>
  </fieldset>
</FormFieldset>
`;

exports[`FormFieldset should render form fieldset required variant 1`] = `
<FormFieldset
  fieldId="id"
  isRequired={true}
  label="label"
>
  <fieldset
    className="pf-c-form__fieldset"
  >
    <label
      className="pf-c-form__label"
      htmlFor="id"
    >
      <span
        className="pf-c-form__label-text"
      >
        label
      </span>
      <span
        aria-hidden="true"
        className="pf-c-form__label-required"
      >
        *
      </span>
    </label>
    <input
      id="input-id"
    />
  </fieldset>
</FormFieldset>
`;

exports[`FormFieldset should render form fieldset variant with node helperText 1`] = `
<FormFieldset
  fieldId="label-id"
  helperText={
    <span>
      Help text!
    </span>
  }
  label="Label"
>
  <fieldset
    className="pf-c-form__fieldset"
  >
    <label
      className="pf-c-form__label"
      htmlFor="label-id"
    >
      <span
        className="pf-c-form__label-text"
      >
        Label
      </span>
    </label>
    <input
      id="input-id"
    />
    <div
      aria-live="polite"
      className="pf-c-form__helper-text"
      id="label-id-helper"
    >
      <span>
        Help text!
      </span>
    </div>
  </fieldset>
</FormFieldset>
`;

exports[`FormFieldset should render form fieldset variant with node label and help text 1`] = `
<FormFieldset
  fieldId="id"
  helperText="this is helper text"
  label={
    <span>
      Label
    </span>
  }
>
  <fieldset
    className="pf-c-form__fieldset"
  >
    <label
      className="pf-c-form__label"
      htmlFor="id"
    >
      <span
        className="pf-c-form__label-text"
      >
        <span>
          Label
        </span>
      </span>
    </label>
    <input
      aria-label="input"
    />
    <div
      aria-live="polite"
      className="pf-c-form__helper-text"
      id="id-helper"
    >
      this is helper text
    </div>
  </fieldset>
</FormFieldset>
`;

exports[`FormFieldset should render inline form fieldset variant 1`] = `
<FormFieldset
  fieldId="label-id"
  isInline={true}
>
  <fieldset
    className="pf-c-form__fieldset pf-m-inline"
  >
    <input
      id="input-id"
    />
  </fieldset>
</FormFieldset>
`;
